### ISSUE TITLE
Possible race condition on 0.16+ causing hang at fd access

### ISSUE THREAD

**Issue by parkerlreed** [sddm/sddm#1226]
Arch Linux ARM Aarch64
Linux alarm 4.9.140+ #1 SMP PREEMPT Thu Jan 2 09:32:08 PST 2020 aarch64 GNU/Linux

When using this commit or anything newer (performed a full bisect) SDDM appears to start normally but hangs at an fd access. https://github.com/sddm/sddm/commit/2187cd880c62ad415ec360493fe3a0ead01bd772

strace log http://ix.io/26CU

Tested all the way up to current git master and face the exact same issue. 

Partial debug output 

```
#0  0x0000007f9dd4df68 in poll () from /usr/lib/libc.so.6
No symbol table info available.
#1  0x0000007f9ce25ff8 in ?? () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#2  0x0000007f9ce2617c in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#3  0x0000007f9e30b95c in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
No symbol table info available.
#4  0x0000007f9e2a06fc in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
No symbol table info available.
#5  0x0000007f9e2aaa04 in QCoreApplication::exec() () from /usr/lib/libQt5Core.so.5
No symbol table info available.
#6  0x000000558c90a248 in main (argc=<optimized out>, argv=0x7ffa78f338) at /usr/src/debug/sddm-0.18.1/src/daemon/DaemonApp.cpp:138
        arguments = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, _M_i = -1}, <No data fields>}}}, 
                  alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x5592babfc0}, d = 0x5592babfc0}}, <No data fields>}
        app = {<QCoreApplication> = {<No data fields>}, static staticMetaObject = {d = {superdata = {direct = 0x7f9e551680 <QCoreApplication::staticMetaObject>}, stringdata = 0x558c975850 <qt_meta_stringdata_SDDM__DaemonApp>, 
              data = 0x558c976410 <qt_meta_data_SDDM__DaemonApp>, static_metacall = 0x558c90b5b8 <SDDM::DaemonApp::qt_metacall(QMetaObject::Call, int, void**)+4294951144>, relatedMetaObjects = 0x0, extradata = 0x0}}, first = true, 
          static self = 0x7ffa78f1a8, m_lastSessionId = 0, m_testing = false, m_displayManager = 0x5592ba3590, m_powerManager = 0x5592bb1960, m_seatManager = 0x5592bb6750, m_signalHandler = 0x5592bbad80}
```

The fd's 14-15 appear to be sockets when checking the hung process

```
[root@alarm sddm]# lsof -a -p 63947
COMMAND   PID USER   FD      TYPE             DEVICE SIZE/OFF   NODE NAME
sddm    63947 root  cwd       DIR              179,2     4096      2 /
sddm    63947 root  rtd       DIR              179,2     4096      2 /
sddm    63947 root  txt       REG              179,2   661576 793257 /usr/bin/sddm
sddm    63947 root  mem       REG              179,2  3031632 811876 /usr/lib/locale/locale-archive
sddm    63947 root  mem       REG              179,2    51152 792973 /usr/lib/libnss_files-2.29.so
sddm    63947 root  mem       REG              179,2   407432 792516 /usr/lib/libpcre.so.1.2.11
sddm    63947 root  mem       REG              179,2 27981552 793108 /usr/lib/libicudata.so.65.1
sddm    63947 root  mem       REG              179,2    84128 792679 /usr/lib/libresolv-2.29.so
sddm    63947 root  mem       REG              179,2    22376 792076 /usr/lib/libkeyutils.so.1.9
sddm    63947 root  mem       REG              179,2    63592 792554 /usr/lib/libkrb5support.so.0.1
sddm    63947 root  mem       REG              179,2    14184 793292 /usr/lib/libcom_err.so.2.1
sddm    63947 root  mem       REG              179,2   223200 792253 /usr/lib/libk5crypto.so.3.1
sddm    63947 root  mem       REG              179,2  1054728 792568 /usr/lib/libkrb5.so.3.3
sddm    63947 root  mem       REG              179,2   141432 793355 /usr/lib/libgpg-error.so.0.27.0
sddm    63947 root  mem       REG              179,2  1322448 792543 /usr/lib/libglib-2.0.so.0.6200.4
sddm    63947 root  mem       REG              179,2   587592 793347 /usr/lib/libzstd.so.1.4.4
sddm    63947 root  mem       REG              179,2   477384 792440 /usr/lib/libpcre2-16.so.0.9.0
sddm    63947 root  mem       REG              179,2  2103296 793088 /usr/lib/libicuuc.so.65.1
sddm    63947 root  mem       REG              179,2  3332352 792597 /usr/lib/libicui18n.so.65.1
sddm    63947 root  mem       REG              179,2    83848 808168 /usr/lib/libdouble-conversion.so.3.1.5
sddm    63947 root  mem       REG              179,2  2866912 792970 /usr/lib/libcrypto.so.1.1
sddm    63947 root  mem       REG              179,2   618192 792685 /usr/lib/libssl.so.1.1
sddm    63947 root  mem       REG              179,2   352264 792854 /usr/lib/libgssapi_krb5.so.2.2
sddm    63947 root  mem       REG              179,2    14288 793079 /usr/lib/libdl-2.29.so
sddm    63947 root  mem       REG              179,2    92048 792448 /usr/lib/libz.so.1.2.11
sddm    63947 root  mem       REG              179,2   838680 793181 /usr/lib/libgcrypt.so.20.2.5
sddm    63947 root  mem       REG              179,2    22616 792745 /usr/lib/libcap.so.2.28
sddm    63947 root  mem       REG              179,2   124736 792535 /usr/lib/liblz4.so.1.9.2
sddm    63947 root  mem       REG              179,2   153528 792372 /usr/lib/liblzma.so.5.2.4
sddm    63947 root  mem       REG              179,2    31016 792294 /usr/lib/librt-2.29.so
sddm    63947 root  mem       REG              179,2   645072 792387 /usr/lib/libm-2.29.so
sddm    63947 root  mem       REG              179,2   342088 792066 /usr/lib/libdbus-1.so.3.19.11
sddm    63947 root  mem       REG              179,2   160648 793010 /usr/lib/libpthread-2.29.so
sddm    63947 root  mem       REG              179,2  1829880 792089 /usr/lib/libc-2.29.so
sddm    63947 root  mem       REG              179,2   747936 792106 /usr/lib/libgcc_s.so.1
sddm    63947 root  mem       REG              179,2 15006776 792811 /usr/lib/libstdc++.so.6.0.25
sddm    63947 root  mem       REG              179,2  5758752 808233 /usr/lib/libQt5Core.so.5.14.0
sddm    63947 root  mem       REG              179,2  2050008 808278 /usr/lib/libQt5Network.so.5.14.0
sddm    63947 root  mem       REG              179,2   840464 793190 /usr/lib/libsystemd.so.0.27.1
sddm    63947 root  mem       REG              179,2  4880984 809110 /usr/lib/libQt5Qml.so.5.14.0
sddm    63947 root  mem       REG              179,2   677792 808238 /usr/lib/libQt5DBus.so.5.14.0
sddm    63947 root  mem       REG              179,2   162616 793294 /usr/lib/ld-2.29.so
sddm    63947 root    0r      CHR                1,3      0t0   1238 /dev/null
sddm    63947 root    1u     unix 0xffffffc0d5ff9400      0t0 108887 type=STREAM
sddm    63947 root    2u     unix 0xffffffc0d5ff9400      0t0 108887 type=STREAM
sddm    63947 root    3u  a_inode               0,10        0   2089 [eventfd]
sddm    63947 root    4r      CHR                1,9      0t0   2093 /dev/urandom
sddm    63947 root    5u     unix 0xffffffc0cec29400      0t0 108893 type=DGRAM
sddm    63947 root    6u  a_inode               0,10        0   2089 [eventfd]
sddm    63947 root    7u     unix 0xffffffc0cec29000      0t0 108894 type=STREAM
sddm    63947 root    8u     unix 0xffffffc0b32b7c00      0t0 109950 type=STREAM
sddm    63947 root    9u     unix 0xffffffc0b32b5800      0t0 109951 type=STREAM
sddm    63947 root   10u     unix 0xffffffc0b32b5000      0t0 109952 type=STREAM
sddm    63947 root   11u     unix 0xffffffc0b32b6c00      0t0 109953 type=STREAM
sddm    63947 root   12u     unix 0xffffffc0b32b4000      0t0 109954 type=STREAM
sddm    63947 root   13u     unix 0xffffffc0b32b7800      0t0 109955 type=STREAM
sddm    63947 root   14u     unix 0xffffffc0b32b5400      0t0 109956 type=STREAM
sddm    63947 root   15u     unix 0xffffffc0b32b7400      0t0 109957 type=STREAM
```

**Comment by Vogtinator** [comment]
I don't think it's actually hanging, it's just in the event loop.

Most likely it's waiting for a seat to appear, until then there's just no greeter.

**Comment by parkerlreed** [comment]
Strange. This only seems to happen on ARM (Aarch64 in this case)

This is the only other mention I can find of it 

https://archlinuxarm.org/forum/viewtopic.php?t=13437

That's what led me to bisecting it.

**Comment by Vogtinator** [comment]
You could try `dbus-monitor --system` to debug this.

**Comment by parkerlreed** [comment]
http://ix.io/2kpW

**Comment by Vogtinator** [comment]
You'll have to start it as root before running sddm, otherwise the relevant activity from/to sddm will be missing.

**Comment by parkerlreed** [comment]
```
systemctl stop sddm
pacman -Syu sddm
```

sddm 0.18.1-2

In another SSH session I captured the dbus-monitor output running as root when doing

```
systemctl start sddm
```

http://ix.io/2kMx

**Comment by Vogtinator** [comment]
```
method call time=1588607735.084616 sender=:1.97 -> destination=org.freedesktop.login1 serial=19 path=/org/freedesktop/login1/seat/seat0; interface=org.freedesktop.DBus.Properties; member=Get
   string "org.freedesktop.login1.Seat"
   string "CanGraphical"
method return time=1588607735.084831 sender=:1.6 -> destination=:1.97 serial=417 reply_serial=19
   variant       boolean false
```

So the seat is discovered properly, but logind says that it is not graphical. This means it's not a bug in sddm, but an interaction of systemd with device drivers.

**Comment by parkerlreed** [comment]
Thank you so much. This is using the Nvidia drivers on a Nintendo Switch (same platform/drivers as the Jetson Nano). Any suggestions on where this could possibly be fixed/patched?

**Comment by Vogtinator** [comment]
I don't know how systemd determines that, but it probably tries to match udev devices to seats and couldn't find any graphical output. A quick guess is that you need something like https://github.com/systemd/systemd/pull/11438/files.

**Comment by parkerlreed** [comment]
Thanks. Will try and give that a shot.

**Comment by parkerlreed** [comment]
Hmm I'll have to figure something out. The only two PCI devices are the internal bridge and Wi-Fi device.

**Comment by parkerlreed** [comment]
Interestingly there doesn't seem to be a session at all when booting without a login manager enabled. Only the SSH I'm using to grab the information. 

```
[parker@alarm ~]$ loginctl list-sessions
SESSION  UID USER   SEAT TTY
      1 1001 parker

1 sessions listed.
```

**Comment by parkerlreed** [comment]
The commit where it broke mentions the old behavior should be the fallback if nothing is detected. Is there any way to force the old behavior on current sddm?

**Comment by parkerlreed** [comment]
So the TX1 seems to have no GPU device path anywhere under /sys. Which leaves system nothing to attach seat0 no, which in turn means its not Graphical. Somehow the old behavior works with this setup... *insert Linus F*** you Nvidia here*

**Comment by parkerlreed** [comment]
Wait a second. It *is* attached

So why is this being seen as non Graphical?

```
[root@alarm parker]# loginctl seat-status seat0
seat0
         Devices:
                  ├─/sys/devices/50000000.host1x/tegradc.0/graphics/fb0
                  │ graphics:fb0 "tegra_fb"
                  ├─/sys/devices/50000000.host1x/tegradc.1/graphics/fb1
                  │ graphics:fb1 "tegra_fb"
                  ├─/sys/devices/7000c500.i2c/i2c-2/2-0049/input/input0
                  │ input:input0 "touchscreen"
                  ├─/sys/devices/7000d000.i2c/i2c-4/4-003c/max77620-onoff/input/input1
                  │ input:input1 "max77620_onoff"
                  ├─/sys/devices/70030000.hda/sound/card0
                  │ sound:card0 "tegrahda"
                  │ └─/sys/devices/70030000.hda/sound/card0/input2
                  │   input:input2 "tegra-hda HDMI/DP,pcm=3"
                  ├─/sys/devices/70090000.xusb/usb1
                  │ usb:usb1
                  ├─/sys/devices/70090000.xusb/usb2
                  │ usb:usb2
                  ├─/sys/devices/backlight/backlight/pwm-backlight
                  │ backlight:pwm-backlight
                  ├─/sys/devices/gpio-keys/input/input4
                  │ input:input4 "gpio-keys"
                  ├─/sys/devices/sdhci-tegra.0/leds/mmc0::
                  │ leds:mmc0::
                  ├─/sys/devices/sound/sound/card1
                  │ sound:card1 "tegrasndt210ref"
                  │ └─/sys/devices/sound/sound/card1/input3
                  │   input:input3 "tegra-snd-t210ref-mobile-rt565x Headphone Jack"
                  └─/sys/devices/virtual/misc/rfkill
                    misc:rfkill
```

**Comment by parkerlreed** [comment]
Finally got this fixed. Simple in the end

```
SUBSYSTEM=="graphics", KERNEL=="fb[0-9]", TAG+="master-of-seat"
```

It doesn't advertise a driver name.

**Comment by ChrisRCarlin** [comment]
A quick note:
I ran into this issue when I updated a couple of Raspberry Pi Ubuntu installations, and in searching I think a lot of other people have been running into it as well. @parkerlreed 's solution worked like a charm, so hopefully this comment can help other people find this fix

To implement the solution create a file /etc/udev/rules.d/99masterseat.rules :
`SUBSYSTEM=="graphics", KERNEL=="fb[0-9]", TAG+="master-of-seat"`

And then to apply it:
`sudo udevadm trigger`

or just reboot

To summarize, sddm was starting, but sat waiting for systemd / logind to tell it there was a screen ready for the greeter. udev uses that master-of-seat tag to say the seat was ready to go, but it wasn't being applied for some reason. 

```
~$ uname -a
Linux hostname 5.4.0-1047-raspi #52-Ubuntu SMP PREEMPT Wed Nov 24 08:16:38 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
```

**Comment by Vogtinator** [comment]
The best way to make systemd use framebuffer devices for `CanGraphical` as well is to use the `nomodeset` kernel option.

However, in the vast majority of cases there should be a DRM device available, if not it's usually some driver issue.

> I ran into this issue when I updated a couple of Raspberry Pi Ubuntu installations, and in searching I think a lot of other people have been running into it as well.

On the RPi, you should have a DRM device available. If not, try to reinstall kernel and firmware. With just fbX you won't have hardware acceleration.
